### DISCUSSION TITLE
Support transitions based on list mutations

### DISCUSSION THREAD

**Discussion by sarahdayan** [tailwindlabs/headlessui#732] [Ideas]
Hey Tailwind UI team! Thanks for the lovely library.

I'm creating ecommerce examples with Tailwind UI and Headless UI, and a typical scenario where you want to animate transitions is when **adding items to or deleting items from a list** (e.g., adding or removing items from a cart).

Here's an example with todos:

https://user-images.githubusercontent.com/5370675/129459738-aefa690f-38f4-453c-a154-12c35670e444.mov

When doing so, you typically want to delete items no differently from if you weren't transitioning, and let the transition component handle the animation. This is currently not possible with Headless UI because it doesn't provide a way to track mutations on a list.

For example, the following React example wouldn't work (items would disappear but without transitions):

```jsx
function App() {
  const [todos, setTodos] = useState([
    'Buy eggs',
    'Pay bills',
    'Invite friends over',
    'Fix the TV',
  ]);

  function onDelete(todoToRemove) {
    setTodos(todos.filter((todo) => todo !== todoToRemove));
  }

  return (
    <ul>
      {todos.map((todo) => (
        <Transition
          leave="transition-opacity duration-150"
          leaveFrom="opacity-100"
          leaveTo="opacity-0"
          as="li"
          key={item}
          show={todos.includes(todo)}
        >
          {todo} <button onClick={() => onDelete(todo)}>Delete</button>
        </Transition>
      ))}
    </ul>
  );
}
```

To make it work, you'd need to mutate some prop on each item (e.g., `isRemoved: true`) instead of mutating the list.

```jsx
function App() {
  const [todos, setTodos] = useState([
    {
      title: 'Buy eggs',
      isRemoved: false,
    },
    // ...
  ]);

  function onDelete(todoToRemove) {
    setTodos(todos.map((todo) => ({ ...todo, isRemoved: todo === todoToRemove })));
  }

  // ...
}
```

This solution works, **but forces you to adapt your data to implementation details of the transition library.** This is a code smell, and can be annoying when you need to perform further manipulation on the data (e.g., computing a total price on a list of items in a cart, extracting the items count) because you must filter out "soft deleted" items.

A possible solution would be to do something similar to what [React Transition Group](http://reactcommunity.org/react-transition-group/) does with its [`TransitionGroup`](http://reactcommunity.org/react-transition-group/transition-group) component, by tracking mutations on the passed `children` and automatically update the `show` prop.

@RobinMalfait Is there already something similar in the Headless UI roadmap? And if not, is this a problem you'd be interested seeing solved in Headless UI?

**Comment by levsoroka** [comment]
Thanks @sarahdayan suggested workaround did help me overcome this same issue that you are describing. And I do agree that, it does become an additional effort to manage additional prop rather than having filtered array of only needed items.

**Comment by jdortegar** [comment]
same request here !

**Comment by eakl** [comment]
Same request

**Comment by zevisert** [comment]
Came looking for [`TransitionGroup`](http://reactcommunity.org/react-transition-group/transition-group) too, but while I'm here.. Another way to handle the soft-delete state is with `setTimeout`, **but that has it's own problems too**: if someone navigates away before the animation is done, or otherwise performs some action, they'll get old state. 

```jsx
function SomeComponent({ onRemove }: { onRemove: MouseEventHandler<HTMLButtonElement> }) {
  const [shown, setShown] = useState(true);
  const delayRemove = (evt) => {
    setShown(false);
    setTimeout(onRemove(evt), 150);
  };

  return (
    <Transition
      show={shown}
      leave='transition duration-150'
      leaveFrom='opacity-100'
      leaveTo='opacity-0'
      ...
    >
      ...
      <button onClick={delayRemove}>Remove</button>
    </Transition>
  ) 
}
```

**Comment by arimendelow** [comment]
Same request here! @RobinMalfait any news on this?

**Comment by nataliemeurer** [comment]
Same request over here! Would love a headlessui transition group.

**Comment by frolic** [comment]
I haven't published this to npm or written proper docs yet, but I have been playing with an `AnimatePresence` port for Headless UI here: https://github.com/holic/headless-ui-transition-group

Would love input!

**Comment by quagliero** [comment]
Another +1 here.

**Comment by arimendelow** [comment]
@RobinMalfait @adamwathan still no updates on this? :(

**Comment by dstrunk** [comment]
Here's my go at this; first, the solution, then a short explanation:

```jsx
function List() {
    const [items, setItems] = useState([]);

    const close = (item) => (e) => {
        setItems(items.filter((i) => i.id !== item.id));
    };

    return (
        {items.length > 0 && (
            <>
                {items.map((item) => (
                    <ListItem item={item} key={item.id} close={close} />
                ))}
            </>
        )}
    );
}

function ListItem({ item, close }) {
    const [show, setShow] = useState(true);

    return (
        <Transition
            appear={true}
            show={show}
            afterLeave={close(item)}
            as={Fragment}
        >
            <Transition.Child
                as={'div'}
                className="w-full max-w-sm overflow-hidden bg-white rounded-lg shadow-lg pointer-events-auto ring-1 ring-black ring-opacity-5"
                enter="transform ease-out duration-300 transition"
                enterFrom="translate-y-2 opacity-0 sm:translate-y-0 sm:translate-x-2"
                enterTo="translate-y-0 opacity-100 sm:translate-x-0"
                leave="transition ease-in duration-100"
                leaveFrom="opacity-100"
                leaveTo="opacity-0"
            >
                {item.name}
                <button onClick={() => setShow(!show)}>Close</button>
            </Transition.Child>
        </Transition>
    )
}
```

First, create a `close` method (as you would normally). This method removes the item selected from the array, nothing special.

Next, separate the mapped list items into their own `ListItem` component. This component passes in individual items as an `item` prop, as well as the `close` method from the parent. Creating a separate component allows for managing visibility for individual list items via a state-managed `show` variable. By default each list item is set to show by default by setting `show` to `true`.

Finally, note the following in the `ListItem`'s `<Transition>` props:

1. `appear={true}` - Ensure any items fade in when initialized.
2. `show={show}` - Use the state-managed `show` variable to control visibility
3. `afterLeave={close(item)}` - The sauce to take care of this request; after the `leave` transition has completed, it will fire this method, which calls the `List` component's `close` method to remove the (now hidden) list item from the `items` array.
4. `<button onClick={() => setShow(!show)}>Close</button>` - Set the `show` visibility.

Hope that helps anybody still wrestling with this!

**Comment by panieldark** [comment]
Still unfortunately unsupported, had to switch to Framer :/

**Comment by jbownzino** [comment]
+1 for addressing this natively.. running into this issue. Thanks to everyone who provided their solutions.

**Comment by jbownzino** [comment]
Heads up, I switched to using a lightweight package named AutoAnimate to solve this -- https://github.com/formkit/auto-animate. Code to implement and bundle size is minimal and works perfectly for this use case.

**Reply by gkkirsch** [reply]
I just tried this package out. I love it. It doesn't play well with headless-ui transitions. Have you run into that?
